egui/src/widgets/text_edit.rs

```diff
@@ -108,6 +109,50 @@ impl CCursorPair {
     }
 }
 
+/// Trait contraining what types [`TextEdit`] may use as
+/// an underlying buffer
+pub trait TextBuffer:
+    AsRef<str> + Into<String> + PartialEq + Clone + Default + Send + Sync + 'static + std::fmt::Display
+{
+    /// Inserts text `text` into this buffer at character index `ch_idx`.
+    ///
+    /// # Notes
+    /// `ch_idx` is a *character index*, not a byte index.
+    ///
+    /// # Return
+    /// Returns how many *characters* were successfully inserted
+    fn insert_text(&mut self, text: &str, ch_idx: usize) -> usize;
+
+    /// Deletes a range of text `ch_range` from this buffer.
+    ///
+    /// # Notes
+    /// `ch_range` is a *character range*, not a byte range.
+    fn delete_text_range(&mut self, ch_range: Range<usize>);
```

Comment:
```suggestion
    fn delete_char_range(&mut self, ch_range: Range<usize>);
```

is a more descriptive name